README.md

```diff
@@ -0,0 +1,396 @@
+Then, it is up to apps to depend on the necessary plugins.
+
+## Programming Errors
+
+Programming errors, such as calling a method at an inappropriate time, should be handled with `panic(..)` instead of errors.
+If an error is returned, it will likely be logged and may not be seen.
+A panic, on the other hand, is very noticeable and carries a stack trace that can help the programmer figure out what they've missed.
+Try to arrange for such panics to happen consistently, so that such programming errors are quick to find.
+
+# Future Plans
+
+## Component Linting
+
+With good detection of components (already used to generate COMPONENTS.md and CODEOWNERS), we can check that the guidelines are followed.
+For example, this check could easily verify that components are not nested, and that every component has a `Component` type and `Module` value.
+
+## Config Reducers
+
+We can use a concept similar to that defined by Redux to simplify the DD configuration used by each component that needs it.
+This would involve a "reducer" that extracts data from Viper (or whatever we switch to) and places it in a component-specific struct.
+Using struct tags and a utility function would allow for a very regular, greppable arrangement of configuration parameters with very little per-component boilerplate.
+This will also ease testing of components: tests can simply provide a filled-in configuration struct, instead of manually setting configuration parameters.
+
+# Open Questions
+
+## Subprocesses
+
+We want to support running some "things" (we should have a term for this!) as subprocesses, as is currently done for trace-agent, system-probe, process-agent, and security-agent.
+Should these be different binaries (as they are now), or the same binary with different arguments?
+
+## Representing Phases with Types
+
+Olivier Gaça suggests that we could avoid warnings in comments to only call certain methods before/after startup by using distinct types to represent the component before and after startup.
+
+## ModuleParams Defaults
+
+Currently, the guide recommends that `ModuleParams` be optional and that its zero value be meaningful.
+But this may result in a lot of `if params.Xyz == "" { params.Xyz = "real default" }`, which could be annoying and ugly.
+If component users were encouraged to use `fx.Replace(foo.ModuleParams{ Xyz: "specific value" })` to specify parameters, then the component's module could use `fx.Supply(ModuleParams{ Xyz: "real default", ... }` to supply nonzero defaults.
+
+## Enabling / Disabling Component Bundles
+
+Consider configs like `logs_enabled` or `apm_config.enabled` -- when these are false, it means that several components that are already instantiated should not actually start.
+One options here is to thread a bunch of `Enable` methods through these components.
+Then an app would call
+```go
+fx.Invoke(func(config config.Component, agent logsagent.Component) {
+    if config.GetBool("logs_enabled") {
+        logsagent.Enable()
```

Comment:
That's a good idea.  This is part fo the "enablement" question that I need to figure out -- do we want logs-agent to start just because `comp/logs/agent.Module` is included in the App?  I want to both avoid things happening "magically" (and perhaps surprisingly, when we realize that logs-agent is running quietly in the trace-agent process!) and avoid too much boilerplate of calling Enable() or passing ModuleParams{Disable: true} all over the place.